### DISCUSSION TITLE
Wireless MPG?

### DISCUSSION THREAD

**Discussion by Team-Deepflights** [grblHAL/core#462] [General]
I finished the mill and would like to use a DRO at it.
I found a wired one, i posted a question about wiring at Phil’s repo.
But I also found one which is wireless.
My thougt is to use an USB port at the laptop running IOsender and send the command over it.
Wrong way or am I right?
I tried to get some information on web but the results are not really useful, most refer on Linux CNC or MACH.

https://www.ebay.de/itm/275222395036?mkcid=16&mkevt=1&mkrid=707-127634-2357-0&ssspo=uopd6Cl1RxC&sssrc=4429486&ssuid=kqkWdZmoTVG&var=&widget_ver=artemis&media=COPY

could someone help me please?

BTW, here is the mill, it use driven ball screws, work very nice.
![image](https://github.com/grblHAL/core/assets/96638334/5432b914-ab3c-4ac8-87f0-e62532bdde59)

**Comment by terjeio** [comment]
> My thougt is to use an USB port at the laptop running IOsender and send the command over it.

If it acts as a keyboard then the ioSender shortcuts can perhaps be modified to match the MPG. But using the wheel and output data to the display might be problematic, that would require custom code. And detailed documentation could be hard to find?

FYI the ioSender keymappings can be saved out from the Settings: App tab, the file is in XML format and can be edited - it will be loaded on startup when available.

Very nice mill BTW!

**Reply by Team-Deepflights** [reply]
Still not sure which way to go, as I only would like to use a pendant, I really havo no need to send DRO information.
Phil suggested to use an X-Box controller, but to be honest, I prefer something that feels like a real controller 😉
As it seems that this is not a plug & play solution, maibe I get a cheaper one and traduce the input as keyboard strokes with an Arduino.

When I have some time, I’ll provide some detailed information about the mill as U-Tube video 😉

**Reply by terjeio** [reply]
There are a couple of options available if connected to the controller, but none are currently wireless:

* [RT jog controller](https://github.com/Expatria-Technologies/RT_Jog_Controller) (I2C)
* [Pane](https://github.com/dresco/grblPANEL)l, not exactly a pendant... Uses ModBus for comms.

If your spindle is not connected via RS485/ModBus (I assume you use one of Phil's boards):
* DYI based on [this project](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack) (or derivatives, see issues/discussions).
* A project by @nickshl that looks good and may soon become available?

**Comment by nickshl** [comment]
> here is the mill

It is not a mill, it is a router 😁
But one company, that often throw their ad to me on youtube, hesitate to call their product "_router_", so they invented term "_gantry mill_" 😅.
Anyway, nice machine!

> * A project by @nickshl that looks good and may soon become available?

I plan to release it by the end of this week... but I planned to release it probably couple month ago, so no promises 😅
And it will be open source project(_software and schematic, except PCB which I plan to sell_) - there only 3 essential components needed to build DIY edition one: BlackPill, display module and MPG encoder, all of them could be purchased relatively cheap from AliExpress.

**Reply by Team-Deepflights** [reply]
Well, i would not call a 350kg „lightweight“ thingy with 250mm. Z axis working height a router 😉
But I admit, as the A-axis is still not installed, it looks like a router, gantry mill or whatever.
I had a look at your site but I did nit find any closer information about it.
The pendant is for me a „nice to have“ but not something i need urgently.
So I look how your release would cover my needs.

Thank you 🙂

**Comment by phil-barrett** [comment]
For what it is worth, [here is a project](https://github.com/phil-barrett/PicoPendant) I am working on. The main goal is to create a Pendant interface for grblHAL. It is too soon to say if I will be selling complete boards but could make some PCBs available if there is interest. First steps are using a wired pendant but with a PicoW, I think WiFi or BT could be possible.

I would love to have other people join me in this effort. If you are interested, please introduce yourself in the [discussion section](https://github.com/phil-barrett/PicoPendant/discussion).

**Reply by Team-Deepflights** [reply]
Well Phil, your PCBs are well designed, I am sure this is going to be something reliable.
My problem is that I used the second serial for the ESP.
I think that sending the commands over the Laptop or directly over WIFI is my only way to go.
The only bad thing is that my working shedule is awfully full and I only come back home every 2 weeks for weekend.

**Reply by phil-barrett** [reply]
It can also use the I2C interface. Though, I haven't started on the code for that.  I believe the Pico/RP2040 can act as a slave I2C device so it should be possible.

**Reply by nickshl** [reply]
IMHO, economically doesn't make sense. Custom **SmartPendant™** (_™ - not really_ 😁) with display will be same money(_or even cheaper_) as to buy off the shelf dumb pendant and adapter card.

**Reply by phil-barrett** [reply]
Link please.

**Reply by nickshl** [reply]
I don't have link... yet. Only prototype... three actually: for my mill, for my lathe and for development. And one more I sent to @terjeio. I also have parts to build two more and three more boards that require to order displays and encoders to finish it.
![IMG_20231128_230917176~2](https://github.com/grblHAL/core/assets/38679154/810c1f37-9852-4071-b301-ec6a3b2eaf61)
As I mention earlier in this discussion, I plan to publish it by the end of this week.

**Reply by phil-barrett** [reply]
It looks interesting though the devil is in the details. Looking forward to seeing them, especially the BOM.

**Reply by nickshl** [reply]
@phil-barrett, I have a question for you: how do you deal with sales tax? You sell some of stuff via Tindie.com, I also considered to sell my board there, however they not charging sales tax to customer. People ask about it years ago, but no change. In case if marketplace doesn't charge sales tax, seller responsible to collect and pay it. And what makes it worse: there no way to set different price depend on destination on Tindie.com which mean one price for all states. But tax rate if depends on state and some states doesn't have it at all. So, I decided to use something else(Etsy) because of that. Just wondering what other people do in this case when they sell on Tindie.com.

**Reply by phil-barrett** [reply]
Yes, Tindie doesn't handle US sales tax. One of their many, many flaws. I currently don't charge customers for it.  I'm only responsible for sales tax to people in my home state and sell only a few to them. If they ever come for me, I will just count up the sales and settle for the tax on that.  For sales to European customers, they are responsible for any VAT/GAT/Duty.  Some countries charge it at point of delivery, some don't.

**Reply by nickshl** [reply]
> I'm only responsible for sales tax to people in my home state and sell only a few to them

I don't think that the case. For example if you sell to Minnesota:
> Remote sellers must collect Minnesota Sales Tax on their sales made in the state, due to the U.S. Supreme Court's decision in the South Dakota v. Wayfair case in 2018.

However there is "_Small Seller Exception_" which doesn't require to collect sales tax if revenue less than $100k and less than 200 retail transaction per 12 month. Many state has something similar: https://www.streamlinedsalestax.org/for-businesses/remote-seller-faqs/remote-seller-state-guidance

I found it just now, at a time when I decided to use Etsy I didn't know about  "_Small Seller Exception_". It so much easier when marketplace handle all of that. That is hilarious that Tindie can't solve it for years...

**Reply by phil-barrett** [reply]
I wanted to circle back to this point now that I understand better.

> IMHO, economically doesn't make sense. Custom SmartPendant™ (™ - not really 😁) with display will be same money(or even cheaper) as to buy off the shelf dumb pendant and adapter card.

For US consumption, an off the shelf dumb pendant is $US 70-75 from Amazon, $US 50 from Aliexpress.  The adaptor board I am working on would be around $US 30.  If you can hit $100 for a complete version of your smartpendant, then it's a very good deal. Though, it's not clear what extras are needed - honestly I am a bit skeptical of the end price. But, it's not all bad. Even at $150, I think there would be demand.  I do see some manufacturing complexity to create a complete unit ready to sell so it's kind of a guess what the final price could be. In terms of competition, the [VistaCNC](https://www.vistacnc.com/a04_purchase/purchase.htm) stuff is a good price check point - the cheapest one with a display is $US 187. Yours has more features so should have a bit more pricing headroom. There are other products worth looking at but I will leave that to you.

Anyway, I think there is plenty of room in the market for a number of different devices and price points.

**Reply by nickshl** [reply]
> For US consumption, an off the shelf dumb pendant is $US 70-75 from Amazon, $US 50 from Aliexpress. The adaptor board I am working on would be around $US 30.

That exactly what I meant: $70 + $30 = $100. The rest of answer is there: https://github.com/grblHAL/core/discussions/465#discussioncomment-8868008

**Comment by dresco** [comment]
My 2p worth..

Am not really a fan of taking logic level UART or I2C signals off board in noisy environments. Obviously it can and does work, but given a choice, I'd rather stick with an industrial standard like Modbus or CAN.

That was my motivation for creating the [Panel](https://github.com/dresco/Plugin_panel) plugin, but perhaps ideal situation would be a standard interface for keypad, mpg, display etc that supported all these transports out of the box?

@Team-Deepflights if you're already using Modbus for your spindle, then the Panel plugin would work with the connections you have (but obviously not wireless). The [reference design](https://github.com/dresco/grblPANEL) might be overkill, but easy enough to build for a BlackPill or similar with a single encoder + some buttons..

**Reply by nickshl** [reply]
I have strong feelings, that many people who said Modbus doesn't know what they talking about.
> [Modbus or MODBUS](https://en.wikipedia.org/wiki/Modbus) is a client/server data [communications protocol](https://en.wikipedia.org/wiki/Communications_protocol) in the [application layer](https://en.wikipedia.org/wiki/Application_layer).

**Communication protocol**. **Application layer**. Very old and very outdated. Basically it is bunch of 16 bit registers. It is work well in industrial application for sensors for example, when all what you need one value from one register. In my opinion it is absolutely not suitable to streaming g-code. Yes, you could do it, but if it can be done doesn't mean it should be done.

 A lot of Modbus robustness comes from physical layer, [RS-485](https://en.wikipedia.org/wiki/RS-485), which is half-duplex differential signaling physical layer. You don't have to use Modbus protocol to use it. But it is not suitable for grbl streaming protocol because it is half-duplex. But grbl is not the only thing in the worlds that need full duplex and robustness of differential signaling, that is why [RS-422](https://en.wikipedia.org/wiki/RS-422) were invented.

If you really think logic level UART won't work for you, just buy two of those modules(_about $10 for both, but you can find similar cheaper_):
![RS422-TTL-two-way-signal-TTL-422-Module-full-duplex](https://github.com/grblHAL/core/assets/38679154/ff1c7dea-8a08-47b1-b15b-738ecb4b8a34)
And install one into pendant and another one as closer as possible to your controller.

There is one more thing that makes Modbus more robust than grblHAL streaming protocol. It is CRC. And there even issue for that: https://github.com/grblHAL/core/issues/440
May be it will be added into grblHAL streaming protocol some day(_but I prefer CRC, not a checksum_).

**Reply by dresco** [reply]
> I have strong feelings, that many people who said Modbus doesn't know what they talking about. <snip>

I assume you're having a dig at me? Yes perhaps I should have said RS485, but would hope that would be obvious in this context. Never mind.. 

I do agree it wouldn't be a good choice for streaming G-code. But we're not talking about that here, at least I'm not. My plugin just passes state information in either Modbus RTU registers, or CAN frames.

> If you really think logic level UART won't work for you, just buy two of those modules

As I said originally, obviously it can be made to work...  'but if it can be done doesn't mean it should be done' ;)

You have your preferences & I have mine, fair enough - plenty of room for both..

**Reply by nickshl** [reply]
> I do agree it wouldn't be a good choice for streaming G-code. But we're not talking about that here, at least I'm not.

My pendant have MicroSD card slot and able to stream simple G-Code programs without connecting PC at all.

> As I said originally, obviously it can be made to work... 'but if it can be done doesn't mean it should be done' ;)

That is why I use logic level UART and it works no problem for me at all on my HiTorque 3990 mini mill with 500W brushless spindle and 4NM closed-loop stepper motors. May be it is too small to cause communication issues.

**Comment by nickshl** [comment]
[Devtronic SmartPendant - a way to control grblHAL based CNC machine without PC](https://github.com/grblHAL/core/discussions/465)

**Comment by terjeio** [comment]
> First steps are using a wired pendant but with a PicoW, I think WiFi or BT could be possible.

BT is now possible, I have [recently changed](https://github.com/grblHAL/core/blob/master/changelog.md#20251130) the code for ESP32, PicoW and HC-05 to allow MPG mode to claim the BT port/stream. The native BT implementations (ESP32/PicoW) are assigned stream instance 20, the HC-05 module whatever regular serial stream instance it is wired to.
BT is also as an option for the passthrough stream of my Pico "port" of the [FluidNC Airedale I/O expander](https://github.com/grblHAL/Plugins_misc/#fluidnc-io-expander) code/protocol. The Pico port is yet to be published, a board should be made for it by someone?
